test/src/unit-cppapi-stats.cc

```diff
@@ -0,0 +1,71 @@
+/**
+ * @file unit-cppapi-stats.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2023-2024 TileDB Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * Tests the C++ API for stats related functions.
+ */
+
+#include "test/support/src/stats.h"
+#include "tiledb/sm/cpp_api/tiledb"
+
+#include <test/support/tdb_catch.h>
+
+using namespace tiledb;
+
+TEST_CASE("stats gathering default", "[stats]") {
```

Comment:
The test does demonstrate a correlation between `is_enabled`, `enable`, and `disable`, though the `enable` and `disable` calls are buried inside of the `ScopedStats`.

I do think it is important to use RAII for the unit test since enabling stats affects the whole process state and could have impact on unit tests which are run afterwards if we leave it in the wrong state.

Hence I don't agree that this tests `ScopedStats` rather than the new APIs themselves.

But I'm happy to put this in the example code too, using RAII to control the stats seems like a best practice anyway.

Reply:
0384f0a